### PR TITLE
Enhance nullable annotations in DataGridViewAddColumnDialog

### PR DESCRIPTION
**Pull request by halgab**
## Proposed changes

- Enhance nullable annotations in `DataGridViewColumnCollectionDialog`, removing almost all `!` in the process

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/10021)

### CODE REVIEW

File: src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/DataGridViewAddColumnDialog.cs [hunk 1]
```diff
@@ -607,78 +609,74 @@ private void DataGridViewAddColumnDialog_Load(object? sender, EventArgs e)
...
-        ITypeDiscoveryService? discoveryService = host.GetService(iTypeDiscoveryServiceType) as ITypeDiscoveryService;
+        ITypeDiscoveryService? discoveryService = host.GetService<ITypeDiscoveryService>();
 
         if (discoveryService is null)
         {
             return;
         }
 
-        ICollection<object> columnTypes = (ICollection<object>)DesignerUtils.FilterGenericTypes(discoveryService.GetTypes(dataGridViewColumnType, false /*excludeGlobalTypes*/));
+        ICollection columnTypes = DesignerUtils.FilterGenericTypes(discoveryService.GetTypes(typeof(DataGridViewColumn), excludeGlobalTypes: false));
```
Comment (halgab) [inline src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/DataGridViewAddColumnDialog.cs:675]: Today `DesignerUtils.FilterGenericTypes` returns an `ArrayList` under the hood. I don't see how this cast to a generic interface could have worked...

File: src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/DataGridViewAddColumnDialog.cs [hunk 2]
```diff
@@ -87,24 +77,24 @@ public DataGridViewAddColumnDialog(DataGridViewColumnCollection dataGridViewColu
     /// </summary>
     private void AddColumn()
     {
-        ComboBoxItem? comboBoxItem = _columnTypesCombo?.SelectedItem as ComboBoxItem;
-        Type? columnType = comboBoxItem?.ColumnType;
+        ComboBoxItem? comboBoxItem = (ComboBoxItem)_columnTypesCombo.SelectedItem!;
```
Comment (dreddy-work) [inline src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/DataGridViewAddColumnDialog.cs:80]: So, you don't see possible `null` warnings on these by analyzer now? i.e `_columnTypesCombo`

Comment (halgab) [reply src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/DataGridViewAddColumnDialog.cs:80]: `_columnTypesCombo` and many other fields can be declared as non null thanks to the addition of the `MemberNotNull` attribute on `InitializeComponent`. This make all those cleanups possible

Comment (dreddy-work) [reply src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/DataGridViewAddColumnDialog.cs:80]: Thank you. I see it now.

Comment (dreddy-work) [reply src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/DataGridViewAddColumnDialog.cs:80]: addign @lonitra  and @JeremyKuhne to highlight similar cases when they review nullability PRs.

Comment (halgab) [reply src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/DataGridViewAddColumnDialog.cs:80]: This file (and the one from #9991) are fairly new additions. I haven't noticed any other file as much impacted as those 2 so far

**Review by lonitra — APPROVED** [review]
LGTM

**Pull request merged**